cmd/controller/main.go

```diff
@@ -89,14 +86,75 @@ func main() {
 		logger.Fatalf("Error building kubeconfig: %v", err)
 	}
 
+	go startPkgController(stopCh, cfg, logger, atomicLevel)
+	go startControllerRuntime(stopCh, cfg, logger, atomicLevel)
+	<-stopCh
+}
+
+func startPkgController(stopCh <-chan struct{}, cfg *rest.Config, logger *zap.SugaredLogger, atomicLevel zap.AtomicLevel) {
+	logger = logger.With(zap.String("controller/impl", "pkg"))
+	logger.Info("Starting the controller")
+
+	const numControllers = 1
+	cfg.QPS = numControllers * rest.DefaultQPS
+	cfg.Burst = numControllers * rest.DefaultBurst
+	opt := reconciler.NewOptionsOrDie(cfg, logger, stopCh)
+
+	kubeInformerFactory := kubeinformers.NewSharedInformerFactory(opt.KubeClientSet, opt.ResyncPeriod)
+	eventingInformerFactory := informers.NewSharedInformerFactory(opt.EventingClientSet, opt.ResyncPeriod)
+
+	subscriptionInformer := eventingInformerFactory.Eventing().V1alpha1().Subscriptions()
+	//deploymentInformer := kubeInformerFactory.Apps().V1().Deployments()
+	//coreServiceInformer := kubeInformerFactory.Core().V1().Services()
+	configMapInformer := kubeInformerFactory.Core().V1().ConfigMaps()
+
+	// Build all of our controllers, with the clients constructed above.
+	// Add new controllers to this array.
+	controllers := []*kncontroller.Impl{
+		subscription.NewController(
+			opt,
+			subscriptionInformer,
+		),
+	}
+	if len(controllers) != numControllers {
+		logger.Fatalf("Number of controllers and QPS settings mismatch: %d != %d", len(controllers), numControllers)
+	}
+
+	// Watch the logging config map and dynamically update logging levels.
+	opt.ConfigMapWatcher.Watch(logging.ConfigMapName(), logging.UpdateLevelFromConfigMap(logger, atomicLevel, logconfig.Controller))
+	// Watch the observability config map and dynamically update metrics exporter.
+	//opt.ConfigMapWatcher.Watch(metrics.ObservabilityConfigName, metrics.UpdateExporterFromConfigMap(component, logger)) // TODO
+	if err := opt.ConfigMapWatcher.Start(stopCh); err != nil {
+		logger.Fatalw("failed to start configuration manager", zap.Error(err))
+	}
+
+	// Start all of the informers and wait for them to sync.
+	logger.Info("Starting informers.")
+	if err := kncontroller.StartInformers(
+		stopCh,
+		subscriptionInformer.Informer(),
+		configMapInformer.Informer(),
+	); err != nil {
+		logger.Fatalf("Failed to start informers: %v", err)
+	}
+
+	// Start all of the controllers.
+	logger.Info("Starting controllers.")
+	go kncontroller.StartAll(stopCh, controllers...)
```

Comment:
Unchecked goroutines make me nervous. I'd expect a failure in any controller to stop the process and all controllers, but based on my reading of https://github.com/knative/pkg/blob/master/controller/controller.go, if a controller crashes, that controller is never restarted and the process continues (after logging the crash via `runtime.HandleCrash()`)

Reply:
https://github.com/kubernetes/apimachinery/blob/b789a6e6b31f49e34cd0b0876791677a85bbb375/pkg/util/runtime/runtime.go#L28-L61

Reply:
it will crash the process.

Reply:
Great, thanks for digging that up!